src/unix/ibmi.c

```diff
@@ -443,9 +446,42 @@ int uv_interface_addresses(uv_interface_address_t** addresses, int* count) {
     }
     address->is_internal = cur->ifa_flags & IFF_LOOPBACK ? 1 : 0;
     if (!address->is_internal) {
-      int rc = get_ibmi_physical_address(address->name, &address->phys_addr);
-      if (rc != 0)
-        r = rc;
+      int rc = -1;
+      size_t name_len = strlen(address->name);
+      /* To get the associated MAC address, we must convert the address to a
+       * line description. Normally, the name field contains the line
+       * description name, but for VLANs it has the VLAN appended with a
+       * period. Since object names can also contain periods and numbers, there
+       * is no way to know if a returned name is for a VLAN or not. eg.
+       * *LIND ETH1.1 and *LIND ETH1, VLAN 1 both have the same name: ETH1.1
+       *
+       * Instead, we apply the same heuristic used by some of the XPF ioctls:
+       * - names > 10 *must* contain a VLAN
+       * - assume names <= 10 do not contain a VLAN and try directly
+       * - if >10 or QDCRLIND returned an error, try to strip off a VLAN
+       *   and try again
+       * - if we still get an error or couldn't find a period, leave the MAC as
+       *   00:00:00:00:00:00
+       */
+      if (name_len <= 10) {
+        /* Assume name does not contain a VLAN ID */
+        rc = get_ibmi_physical_address(address->name, &address->phys_addr);
+      }
+
+      if (name_len > 10 || rc != 0) {
+        /* The interface name must contain a VLAN ID suffix. Attempt to strip
+         * it off so we can get the line description to pass to QDCRLIND.
+         */
+        char* temp_name = uv__strdup(address->name);
+        char* dot = strrchr(temp_name, '.');
+        if (dot != NULL) {
+          *dot = '\0';
+          if (strlen(temp_name) <= 10) {
+            rc = get_ibmi_physical_address(temp_name, &address->phys_addr);
+          }
+        }
+        uv__free(temp_name);
+      }
```

Comment:
Should `r` be set somewhere in this block?

Reply:
I wasn't sure what the behavior was for the return code (the docs didn't seem to say).

The names are coming from the kernel, so the only way there _should_ be an error here is if between the call to Qp2getifaddrs and us calling QDCRLIND the interface was deleted, in which case the MAC address would be set to 00:00:00:00:00:00. Should an error still be returned in such a scenario? I do see now that the AIX code handles any error during the loop as fatal and closes out everything and returns an error, but the Solaris code calls `uv__set_phys_addr` to get the MAC address but ignores any error returned.

Reply:
No, that's fine to ignore errors if it's being done deliberately. That does mean that `r` is never any value other than the initialized `0` -- optionally `r` could be removed and the `return r` at the end of this function changed to `return 0`.

Reply:
Yeah, I noticed after you asked about r that I am assigning rc from get_ibmi_physical_address, but we don't check the result at all, which is pointless.

While fixing this issue, I noticed a bunch of other things that could be tidied up in here, but wanted to keep this PR clean. If I find some time to do a cleanup PR I'll add the suggestion to the list.